### ISSUE TITLE
overlapping edges / flat triangles (solved)

### ISSUE THREAD

**Issue by all4one1** [artem-ogre/CDT#166]
I had a problem, I have found a solution for myself. Just want to share it, maybe it will be somehow useful 
Sorry if it's a duplicate or there is a simpler way to resolve it

Let's say we have 4 points: (0,0), (1,0), (1,1) and (0.5, 0)
For sure, there should be 2 triangles here: 032 and 312
<img width="395" alt="tr" src="https://github.com/artem-ogre/CDT/assets/26969220/a7d5c8f6-7769-4527-aee7-c3d1b770330d">

I just used the simple code for a convex domain

    cdt.insertVertices(
        points.begin(),
        points.end(),
        [](const Point& p) { return p.data[0]; },
        [](const Point& p) { return p.data[1]; }
    );

    cdt.eraseSuperTriangle();

However, I had an extra triangle 031, which was completely flat. 
I found there is the function which checks whenever a point lies to the right from a line, to the left or on a line:
`classifyOrientation(orient2D(p, v1, v2), orientationTolerance);`

**The dafault value of orientationTolerance = 0.** So, it means a point must be located on a line with an exclusive accuracy. 
I change this tolerance to a small value and it worked, i.e. this flat triangle was not generated anymore.

Of course, this is a simple example but I had the same problem with more complex domains. In fact, I tried to implement some simple refinement, where we add a midpoint to our edges (and then repeat triangulation) and came across this behaviour. 

I only don't know how to adequately set orientationTolerance without changing the code of CDT library.

**Comment by all4one1** [comment]
Well, it might be. I cannot check which options I had 3 months ago. 
I have checked points above (and also the case in #172) and now I don't have flat triangles (i.e. in cases if points are indeed exactly on a line). 

However, I believe this problem might have a pure mathematical nature and not a compiler issue. 
Below is there an example

```cpp
#include <iostream>
using namespace std;

#include "CDT/include/CDT.h"

struct Point
{
	double data[2];
	Point(double x_, double y_) : data{ x_, y_ } {};
};

int main()
{
	vector<Point> points;
	points.emplace_back(0.0, 0.0);
	points.emplace_back(1.0, 0.0);
	points.emplace_back(1.0, 1.0);

	double eps = 1e-20;
	points.emplace_back(0.5, 0.0 + eps);


	CDT::Triangulation<double> cdt;
	cdt.insertVertices(
		points.begin(),
		points.end(),
		[](const Point& p) { return p.data[0]; },
		[](const Point& p) { return p.data[1]; }
	);
	cdt.eraseSuperTriangle();

	auto triangles = cdt.triangles;
	cout << "number of triangles = " << triangles.size() << endl;

	for (size_t i = 0; i < triangles.size(); i++)
	{
		cout << triangles[i].vertices[0] << " " << triangles[i].vertices[1] << " " << triangles[i].vertices[2] << endl;
	}
}
```

4 points: (0,0), (1,0), (1,1) and (0.5, 0) must have 2 triangles. Points (0,0), (1,0) and (0.5, 0) are collinear, for sure. 
But imagine point (0.5, 0) has a small disturbance or imperfection and slightly shifted vertically by value of eps, i.e. (0.5, eps). 
There might be a lot of reasons why it might happen. For example, it may be the result of our numerical computations and not just the rounding of floating point operations. 

If eps > 0, we formally have one more triangle even if it's very tiny and flat. But probably we say in this case 'bro, it's insignificant, there still should be 2 triangles'. It is even a little bit philosophical question when we define this threshold of insignificance. But maybe it is a good idea to give a user to define this tolerance. You have already implemented it, look at CDTUtils.hpp file:

```cpp
template <typename T>
PtLineLocation::Enum locatePointLine(
    const V2d<T>& p,
    const V2d<T>& v1,
    const V2d<T>& v2,
    const T orientationTolerance)
{
    return classifyOrientation(orient2D(p, v1, v2), orientationTolerance);
}

template <typename T>
PtLineLocation::Enum
classifyOrientation(const T orientation, const T orientationTolerance)
{
    if(orientation < -orientationTolerance)
        return PtLineLocation::Right;
    if(orientation > orientationTolerance)
        return PtLineLocation::Left;
    return PtLineLocation::OnLine;
}

```
orientationTolerance  = 0 by default. Setting this variable, we can reduce the requirement to be collinear.
I only don't know how to properly set orientationTolerance  without changing the code of CDT lib.

And I noticed, if I set orientationTolerance quite large, the whole triangulation procedure may fail.

**Comment by all4one1** [comment]
the above example with eps = 1e-20 generates 2 triangles, setting orientationTolerance = 1e-10 (for example) fixes the problem

**Comment by artem-ogre** [comment]
Thank you for taking your time and explaining the situation in more detail.
I think I understand the situation. Here is my view on it.

You are describing _"cap triangle"_ issue, a particular case of a degenerate triangle. Dealing with such triangles is done using mesh simplification (mesh cleanup) algorithms and is not a task to be performed by the triangulation algorithm. Triangulation is using robust predicates to guarantee no **truly** degenerate triangles (all points exactly on the same line) will be created. This allows CDT to provide strong guarantee of producing correct result for any possible valid input. Adding tolerances to orientation predicates is opening Pandora box: we can't any longer guarantee correct results for any input.

CDT has `minDistToConstraintEdge` tolerance:
https://artem-ogre.github.io/CDT/classCDT_1_1Triangulation.html#ac5552dc7ad5b1dcc5a22f0bcfeb12c03
But it is used for snapping points to the edges when inserting constraints.

Here are some references to fixing degenerate triangles:
- http://www.cad.zju.edu.cn/home/zhx/papers/ftz_cut_paste.pdf
- https://www.graphics.rwth-aachen.de/media/papers/slicing1.pdf
- https://www.graphics.rwth-aachen.de/media/openmesh_static/Documentations/OpenMesh-6.2-Documentation/a00034.html
- https://doc.cgal.org/latest/Polygon_mesh_processing/group__PMP__geometric__repair__grp.html#ga48008d2b66de8a68a7068f29db15dad6

**Comment by all4one1** [comment]
> Dealing with such triangles is done using mesh simplification (mesh cleanup) algorithms and is not a task to be performed by the triangulation algorithm.

I see, it makes sense. I played with configurations and parameters a litte bit and can confirm it really looks like opening Pandora box when I set a large orientationTolerance. 

However, as long as the setting orientationTolerance is already implemented, I think it might be helpful anyway if we use it very carefully. :) At least, for dealing with rounding of floats.

Thanks for references.
